### DISCUSSION TITLE
Framing effect for table rows

### DISCUSSION THREAD

**Discussion by vvv** [emilk/egui#3069] [Q&A]
_Related: https://github.com/emilk/egui/pull/2904#issuecomment-1513099627, #2000_

I am trying to implement a framing effect for table rows. The desired UI would look like this:

https://github.com/emilk/egui/assets/17462/e8711a62-a99e-47a3-85d8-175260b7365c

(When a row is hovered over, its background color changes to light grey.)

I understand how to frame a button:

1. determine the size,
2. allocate,
3. interact,
4. [paint the background](https://github.com/emilk/egui/blob/860dac69dac6f1378e8556ccd7d39afb98a572e9/crates/egui/src/widgets/button.rs#L199-L200),
5. paint the rest of the 🦉 button **over** the background.

But table row is different. A row made of cells. The row's rectangle is not known until all those cells are painted. And once they are painted, it is already too late to paint the row's background — we would be painting over the cells!

<details>

Like on this demo here, where I'm [painting over](https://github.com/vvv/egui-table-click/blob/f3b71e08950d0b32565f3a04b621a467aa6df4be/src/lib.rs#L131-L132) the middle cell:

https://github.com/emilk/egui/assets/17462/ca072e83-2fa1-48aa-98d1-757cd6a3fb4c

</details>

**Your ideas on how to frame a table row are very welcome!**

**Comment by vvv** [comment]
> The row's rectangle is not known until all those cells are painted. And once they are painted, it is already too late to paint the row's background

This problem would be solved by a two-pass mechanism, proposed in #843.

**Comment by vvv** [comment]
I was able to achieve table row framing effect, though in a [rather hacky manner](https://github.com/emilk/egui/compare/master...vvv:egui:table-row-framing).

### Demo

https://github.com/emilk/egui/assets/17462/0fde5938-7637-4148-bf2f-742b46edfb7b

### Source code

- [Modified `egui`](https://github.com/emilk/egui/compare/master...vvv:egui:table-row-framing)
- [Demo app](https://github.com/vvv/egui-table-click/tree/table-row-framing)

### How the hack works

(Refresher: a row is made of cells, and cells are added with [`TableRow::col`](https://docs.rs/egui_extras/0.22.0/egui_extras/struct.TableRow.html#method.col).)

I've changed the implementation of `TableRow::col` to [update the current row's `Rect`](https://github.com/vvv/egui/blob/76534c4aa5de530c67e87098e8c3bfce495160cd/crates/egui_extras/src/table.rs#L1195) and, if this cell has been interacted with (e.g. hovered), to [remember the `Response`](https://github.com/vvv/egui/blob/76534c4aa5de530c67e87098e8c3bfce495160cd/crates/egui_extras/src/table.rs#L1196-L1204). This data is passed via [`TableBody::this_frame_interaction`](https://github.com/vvv/egui/blob/76534c4aa5de530c67e87098e8c3bfce495160cd/crates/egui_extras/src/table.rs#L773-L774) and [stored](https://github.com/vvv/egui/blob/76534c4aa5de530c67e87098e8c3bfce495160cd/crates/egui_extras/src/table.rs#L609-L612) in the `Context`'s [`Memory::data`](https://docs.rs/egui/0.22.0/egui/struct.Memory.html#structfield.data).

I may not paint the rectangle (`Rect`) of the hovered row after its cells are painted — if I did that, I would be painting **over the cells**. So instead I'm painting the stored rectangle **in the next frame**, as soon as the corresponding row is reached (if ever). I [take](https://github.com/vvv/egui/blob/76534c4aa5de530c67e87098e8c3bfce495160cd/crates/egui_extras/src/table.rs#L587-L591) the rectangle and interaction data ([`TableInteraction`](https://github.com/vvv/egui/blob/76534c4aa5de530c67e87098e8c3bfce495160cd/crates/egui_extras/src/table.rs#L730-L741)) from `Context`'s “data stash” and [paint the frame](https://github.com/vvv/egui/blob/76534c4aa5de530c67e87098e8c3bfce495160cd/crates/egui_extras/src/table.rs#L905-L922).

**Reply by vvv** [reply]
@emilk What do you think about this approach?

**Comment by vvv** [comment]
### Next steps

- [x] ~~As an experiment, I'll move the table row framing logic from `egui_extras` library to my application's code. It should be possible to do so.~~ (Note, that if we conclude that table row framing is a useful feature, we'll want it implemented in `egui_extras`.)
- [x] Then I'll play with row selection. The effect I'm aiming for is shown below. Note that for rows to be clickable, changes of `egui_extras` (similar to those introduced in #2904) are required.

https://github.com/emilk/egui/assets/17462/c801f201-4f3c-4046-99a5-ac5adc343f93

**Comment by vvv** [comment]
> As an experiment, I'll move the table row framing logic from egui_extras library to my application's code. It should be possible to do so.

Nah, I was unable to do that.

```
error[E0502]: cannot borrow `ui` as immutable because it is also borrowed as mutable
  --> src/lib.rs:71:19
   |
54 |         TableBuilder::new(ui)
   |                           -- mutable borrow occurs here
...
71 |             .body(|body| {
   |              ---- ^^^^^^ immutable borrow occurs here
   |              |
   |              mutable borrow later used by call
...
79 |                         highlight_table_row(self.prev_frame_interaction.take().unwrap(), &ui);
   |                                                                                           -- second borrow occurs due to use of `ui` in closure

For more information about this error, try `rustc --explain E0502`.
```

To highlight a rectangle (`Rect`), I need to have a `&Ui` (in order to make [`Ui::painter`](https://docs.rs/egui/0.22.0/egui/struct.Ui.html#method.painter) call). `TableBody` already has a mutable reference to `Ui` (see the diagram below). I [cannot get](https://doc.rust-lang.org/book/ch04-02-references-and-borrowing.html#the-rules-of-references) another reference to `Ui` while `TableBody` is alive.

![230608_egui-table-ui excalidraw](https://github.com/emilk/egui/assets/17462/2291d218-697b-4639-a712-c641078cd8a5)

**Comment by vvv** [comment]
PR #3105 extends egui table with
- framing effect (hovered row is highlighted)
- selectable rows
- configurable cell sensitivity

### Demo

https://github.com/vvv/egui-table-click/assets/17462/b3f78eae-610b-4d01-8028-e8ca3414c4ce

**Comment by emilk** [comment]
* Feature merged in https://github.com/emilk/egui/pull/3347
